sumpy/p2e.py

```diff
@@ -57,16 +57,23 @@ def __init__(self, ctx, expansion, kernels=None,
           number of strength arrays that need to be passed.
           Default: all kernels use the same strength.
         """
-        if kernels is None:
-            kernels = [expansion.kernel]
+        from sumpy.kernel import TargetDerivativeRemover, SourceDerivativeRemover
+        tdr = TargetDerivativeRemover()
+        sdr = SourceDerivativeRemover()
+        expansion = expansion.with_kernel(tdr(expansion.kernel))
+        expansion = expansion.with_kernel(sdr(expansion.kernel))
 
-        KernelComputation.__init__(self, ctx=ctx, kernels=kernels,
-                strength_usage=strength_usage, value_dtypes=None,
-                name=name, device=device)
+        if kernels is None:
+            kernels = [tdr(expansion.kernel)]
+        else:
+            kernels = kernels
+            for knl in kernels:
+                assert tdr(knl) == knl
+                assert sdr(knl) == expansion.kernel
```

Comment:
What is `kernels` for? Can't we supply kernels at expansion time (rather than when calling the constructor)?

Reply:
Yes, kernels are only needed at expansion time and is unnecessary in the constructor, but the way that caching is set up, we need them in the constructor. Overhauling it is a good thing but we would need some unrelated work here in sumpy and in pytential